config-bot/main

```diff
@@ -0,0 +1,373 @@
+
+            # is this a new build?
+            if latest_build_id == last_build_ids.get(stream):
+                continue
+
+            synced = await sync_one_build_lockfile(base_url, method, stream,
+                                                   latest_build_id)
+            if synced:
+                last_build_ids[stream] = latest_build_id
+
+
+async def http_fetch(url):
+    async with aiohttp.request(url=url, method='GET') as resp:
+        if resp.status != 200:
+            eprint(f"Error fetching {url}: got {resp.status}")
+            return None
+        return await resp.read()
+
+
+async def sync_one_build_lockfile(base_url, method, stream, build_id):
+    # we only support direct git pushes for now
+    assert method == 'push'
+
+    # XXX: update for new multi-arch build layout when FCOS is migrated
+    build_dir = f'{base_url}/{stream}/builds/{build_id}'
+    lockfile = 'manifest-lock.generated.x86_64.json'
+    # XXX: switch to `{lockfile}` once this is merged:
+    # https://github.com/coreos/coreos-assembler/pull/593
+    data = await http_fetch(f'{build_dir}/manifest-lock.generated.json')
+    if data is None:
+        # got an error while fetching, just leave and we'll retry later
+        # should probably use a custom Exception for this
+        return False
+
+    async with git:
+        git.checkout(stream)
+        with open(git.path(lockfile), 'wb') as f:
+            f.write(data)
+
+        if git.has_diff():
+            git.commit(f"lockfiles: import from build {build_id}", [lockfile])
+            try:
+                git.push(stream)
+            except Exception as e:
+                # this can happen if we raced against someone/something
+                # else pushing to the ref and we're out of date; we'll retry
+                print(f"Got exception during push: {e}")
+                return False
+
+    return True
+
+
+async def promote_lockfiles(cfg):
+    # this is the only mode we support right now
+    assert cfg['trigger']['mode'] == 'periodic'
+    period = period_to_seconds(cfg['trigger']['period'])
+
+    # we only support direct git pushes for now
+    assert cfg['method'] == 'push'
+
+    source_ref, target_ref = (cfg['source-ref'], cfg['target-ref'])
+
+    last_source_ref_checksum = None
+
+    first = True
+    while True:
+
+        if not first:
+            await asyncio.sleep(period)
+        first = False
+
+        async with git:
+            # is there a new commit?
+            source_ref_checksum = git.rev_parse(source_ref)
+            if last_source_ref_checksum == source_ref_checksum:
+                continue
+
+            git.checkout(target_ref)
+
+            # get the list of lockfiles from the source ref
+            all_files = git.cmd_output('ls-tree', source_ref,
+                                       '--name-only').splitlines()
+            locks = [f for f in all_files
+                     if fnmatch.fnmatch(f, 'manifest-lock.generated.*.json')]
```

Comment:
Minor: could use the `matches_patterns` helper.

Reply:
Done!